[PIR] [Test] Add type generators

effectfully opened the pull request:
This is the types part of QuviQ's [PIR generators PR](https://github.com/input-output-hk/plutus/pull/4559). It's _almost_ ready, I just want to tweak generation of builtins once more (QuickCheck tricked me once again).

effectfully commented:
I was afraid we would generate too big built-in types, but somehow we get exactly the desired kind of output without any tweaking. Here's the biggest output I was able to get with a nearly maximal size parameter: 

```
-- >>> import PlutusCore
-- >>> import Test.QuickCheck
-- >>> sample $ scale (* 3) $ genBuiltinTypeOf $ Type ()
-- Just (SomeTypeIn DefaultUniBool)
-- Just (SomeTypeIn DefaultUniInteger)
-- Just (SomeTypeIn DefaultUniString)
-- Just (SomeTypeIn DefaultUniUnit)
-- Just (SomeTypeIn DefaultUniUnit)
-- Just (SomeTypeIn DefaultUniByteString)
-- Just (SomeTypeIn (DefaultUniApply (DefaultUniApply DefaultUniProtoPair DefaultUniBool) DefaultUniBool))
-- Just (SomeTypeIn (DefaultUniApply (DefaultUniApply DefaultUniProtoPair DefaultUniBool) DefaultUniByteString))
-- Just (SomeTypeIn (DefaultUniApply (DefaultUniApply DefaultUniProtoPair DefaultUniInteger) (DefaultUniApply (DefaultUniApply DefaultUniProtoPair DefaultUniString) DefaultUniInteger)))
-- Just (SomeTypeIn DefaultUniInteger)
-- Just (SomeTypeIn DefaultUniInteger)
```

michaelpj commented on plutus-core/plutus-ir/test/prettyprintingReadable/basic.golden:
```diff
@@ -1 +1 @@
-/\a -> \(x : a) -> x
\ No newline at end of file
+/\a_0 -> \(x_1 : a_0) -> x_1
```
hmmm, should this change go in separately? not a big deal

effectfully replied:
Counterexamples were completely useless without it.

michaelpj replied:
I just mean you could PR it first to make this smaller, but nbd

effectfully replied:
It's a pretty tiny change in an otherwise 2k lines PR, so I'm not gonna bother.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Builtin.hs:
```diff
@@ -0,0 +1,144 @@
+    shrink = error "implement me"
+
+-- | Same as 'Arbitrary' but specifically for Plutus built-in types, so that we are not tied to
+-- the default implementation of the methods for a built-in type.
```
okay, then I guess we should have justifications in each of the cases where we deviate from the existing defintion?

effectfully replied:
Done for `Integer`. I don't know if there's any default handling of `Text` or `ByteString`.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Builtin.hs:
```diff
@@ -0,0 +1,144 @@
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module PlutusCore.Generators.PIR.Builtin where
```
This isn't PIR-specific, right?

effectfully replied:
Yup, I'll move it elsewhere.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR.hs:
```diff
@@ -0,0 +1,9 @@
+module PlutusCore.Generators.PIR (module X) where
```
`PlutusCore.Generators.PIR` :thinking: 
`PlutusIR.Generators`?

michaelpj replied:
Or the type stuff should be in `PlutusCore` but the term stuff should be in `PlutusIR`. Otherwise this is a very weird module structure.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Builtin.hs:
```diff
@@ -0,0 +1,144 @@
+           _ -> []
+
+-- | Convert a @MaybeSomeTypeOf kind@ to a @Maybe (SomeTypeIn DefaultUni)@ and forget the @kind@
+-- (which was only used for generating well-kinded built-in type applications, which we have to do
```
sounds like this should maybe be a standalone Note instead

michaelpj replied:
sounds like this should maybe be a standalone Note instead

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Builtin.hs:
```diff
@@ -0,0 +1,144 @@
+genBuiltinTypeOf :: Kind () -> Gen (Maybe (SomeTypeIn DefaultUni))
+genBuiltinTypeOf kind =
+    withKnownKind kind $ \(_ :: Proxy kind) ->
+        fromMaybeSomeTypeOf <$> arbitrary @(MaybeSomeTypeOf kind)
```
:exploding_head:

michaelpj replied:
No, it's cool!

effectfully replied:
Now that there's a chance we can get away with not having any of this intrinsic typing by having a type-level and a term-level universe, I'm starting to hate all this likely unnecessary complexity.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Common.hs:
```diff
@@ -0,0 +1,74 @@
+import PlutusIR
+import PlutusIR.Error
+
+instance Testable (Either String ()) where
```
should this go upstream?

effectfully replied:
It baffles me that something like this isn't the default way of doing stuff with QuickCheck.

michaelpj replied:
:shrug:

effectfully replied:
In any case, I don't think we can get it upstream. Particularly since `Either String ()` is kinda way too specific and

```haskell
instance Show e => Testable (Either e a)
```

would add quotes for `String`.

I just wanted to make it usable, I propose to leave such tiny unimportant things as is, so that we don't spend another month discussing them.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Common.hs:
```diff
@@ -0,0 +1,74 @@
+    instance PrettyDefaultBy config (NonNegative i) => PrettyBy config (NonNegative i)
+
+-- | Extract all @a_i@ from @a_0 -> a_1 -> ... -> r@.
+argsKind :: Kind ann -> [Kind ann]
```
could go next to `Kind`, it's generically useful

effectfully replied:
> could go next to `Kind`, it's generically useful

I'll do that and 

> I have something like this for Type also in the PIR compiler which should probably go in the main module

I'll let you sort out PIR compiler stuff yourself separately.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+import PlutusCore.Generators.PIR.Common
+import PlutusCore.Generators.PIR.Utils
+
+instance MonadReader r m => MonadReader r (GenT m) where
```
upstream?

effectfully replied:
I think they intentionally don't depend on `mtl` in `quickcheck-transformer`? It's hard to imagine they forgot all such instances by accident. If anything, we should just create a new package with these instances. I checked, it's [not](https://packdeps.haskellers.com/reverse/quickcheck-transformer) on hackage.

michaelpj replied:
... maybe just forget about it :sweat_smile: also upstream hasn't updated for years and is in darcs :grimacing:

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+data GenEnv = GenEnv
+  { geSize               :: Int
+  -- ^ Generator size bound. See Note [Recursion Control and geSize]
+  , geDatas              :: Map TyName (Datatype TyName Name DefaultUni ())
```
why does only `TypeCtx` get an alias?

effectfully replied:
There's `TypeSub` too. Dunno, I didn't find aliasing this one too important, but then I didn't review the terms part, which is where this one is used. I'm going to postpone it until I start dealing with terms.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+  -- ^ Are we currently running in debug-mode (to debug our generators)
+  }
+
+{- Note [Recursion Control and geSize]
```
:+1:

effectfully replied:
Yeah, that one by Maximilian was super helpful.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+          }
+    flip runReader env <$> runGenT g
+
+suchThatMap :: Monad m => GenT m a -> (a -> Maybe b) -> GenT m b
```
haddock?

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+suchThatMap :: Monad m => GenT m a -> (a -> Maybe b) -> GenT m b
+gen `suchThatMap` f = fmap fromJust $ fmap f gen `suchThat` isJust
+
+deliver :: Monad m => GenT m (Maybe a) -> GenT m a
```
haddock?

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:
```diff
@@ -64,6 +70,18 @@ data Kind ann
+toPatFuncKind :: Kind () -> Kind ()
+toPatFuncKind k = KindArrow () (KindArrow () k (Type ())) (KindArrow () k (Type ()))
+
+fromPatFuncKind :: Kind () -> Maybe (Kind ())
```
That `k` appears twice there, which makes it awkward to make it a pattern synonym.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Subst.hs:
```diff
@@ -145,6 +148,64 @@ termSubstFreeNames
 termSubstFreeNames = purely termSubstFreeNamesA
 
+-- | Substitute a closed 'Type' for a type variable in a 'Type'. Does not descend under binders that
+-- bind the same variable as the one we're substituting for. The resulting 'Term' may and likely
```
> Is this behaviour... weird? 

No? I thought it was the perfectly reasonable behavior. As in, the function does exactly the right thing. Do you suspect something?

michaelpj replied:
I was confused. Of course this means the same as "only substitute free names", which is perfectly reasonable. Maybe write that instead?

effectfully replied:
Not "names" -- "name". Changed the comment to

```
-- | Substitute the given closed 'Type' for the given type variable in the given 'Type'. Does not
-- descend under binders that bind the same variable as the one we're substituting for (since from
-- there that variable is no longer free). The resulting 'Term' may and likely will not satisfy
-- global uniqueness.
```

effectfully commented on plutus-core/plutus-ir/test/GeneratorSpec/Substitution.hs:
```diff
@@ -0,0 +1,101 @@
+  letCE "nty1" (normalizeTy ty1)           $ \ nty1 ->
+  letCE "nty2" (normalizeTy ty2)           $ \ nty2 ->
+  letCE "res" (unifyType ctx (Set.fromList $ take nSub xVars) ty1 ty2) $ \ res ->
+  isRight res ==>
+  let sub = fromRight (error "impossible") res
```
No, we want to get an error if the unexpected thing happens. `discard` would just swallow it. It's perfectly fine to get an exception in tests.

michaelpj commented:
Ugh sorry, for some reason half my comments went into a review and half didn't? I'm still going.

effectfully commented on plutus-core/plutus-ir/test/GeneratorSpec/Substitution.hs:
```diff
@@ -0,0 +1,101 @@
+  tabulate "sizes" [show $ min (Set.size $ setOf ftvTy nty1) (Set.size $ setOf ftvTy nty2)] $
+  foldr (.&&.) (property $ nsty1 == nsty2) (map checkSub (Map.toList sub))
+  where
+    allTheVarsCalledX = [ TyName $ Name (fromString $ "x" ++ show i) (toEnum i) | i <- [1..] ]
```
> how is this safe???

Made it

```haskell
  let xVars = take n [TyName $ Name (fromString $ "x" ++ show i) (toEnum i) | i <- [1..]] in
  -- Just for displaying @xVars@ in case of error.
  letCE "xVars" xVars                      $ \ _ ->
```

so that it's clear we're not actually dealing with infinite lists here.

> I am fundamentally still confused about what we're doing with uniques in this PR thinking is there a note somewhere?

See PLT-806. At this point it's very hard to change the current alignment to our way of handling uniques.

michaelpj replied:
yeah, I think it would just be nice to have a note in the generators code that explains that we're doing things a bit weirdly

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Utils.hs:
```diff
@@ -0,0 +1,75 @@
+import PlutusIR.Core.Instance.Pretty.Readable
+
+-- | Show a `Doc` when a property fails.
+ceDoc :: Testable t => Doc ann -> t -> Property
```
So, hedgehog has these sorts of functions, maybe we should copy their naming? https://hackage.haskell.org/package/hedgehog-0.6.1/docs/Hedgehog-Internal-Property.html#v:annotate

effectfully replied:
> So, hedgehog has these sorts of functions, maybe we should copy their naming? https://hackage.haskell.org/package/hedgehog-0.6.1/docs/Hedgehog-Internal-Property.html#v:annotate

The types are wildly different (like, take a different number of arguments and do a different thing), so I'm gonna bother trying to establish a connection.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Utils.hs:
```diff
@@ -0,0 +1,75 @@
+-- * Dealing with fresh names
+
+-- | Freshen a TyName so that it does not equal any of the names in the set.
+freshenTyName :: Set TyName -> TyName -> TyName
```
Shouldn't this be with the other freshening functions?

effectfully replied:
Renamed the function to `freshenTyNameWith` and added the following comment:

```
-- Note that this doesn't have any relation to 'freshenTyName'. Both functions change the unique,
-- but do it in incomparably different ways.
```

One runs in `Quote`, the other takes a set, inserts `0` and the original name into it... It's quite complicated, hence no reason to compare them and store them next to each other.

effectfully commented on plutus-core/plutus-ir/test/GeneratorSpec/Substitution.hs:
```diff
@@ -0,0 +1,101 @@
+  forAllDoc "ty" (genKindAndTypeWithCtx ctx) (shrinkKindAndType ctx) $ \ (k, ty) ->
+  forAllDoc "sub" (genSubst ctx) (shrinkSubst ctx) $ \ sub ->
+  letCE "res" (substType sub ty) $ \ res -> do
+    -- TODO: be more precise.
```
There was a too general error message, I fixed it, but forgot to remove the `TODO`. I'll remove it.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs:
```diff
@@ -0,0 +1,312 @@
+    , "\n\nand\n\n  "
+    , display y
+    , "\n\n"
+    ]
```
Why not use `vcat` to build the combined doc and then `display` it at the end? Simpler than inserting all those newlines manually?

effectfully replied:
I want a specific number of newlines and I'd rather write them out manually than go fishing in the docs of `prettyprinter` in order to make them `hardline`s and trying to understand if it's going to collapse them or something.

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs:
```diff
@@ -0,0 +1,312 @@
+
+The only remaining case is when one of the sides is a type variable (or both). There's a number of
+checks that we perform in order to ensure that it's fine to resolve the variable as the given type,
+those are documented alongside the code performing the checks.
```
good note

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs:
```diff
@@ -0,0 +1,312 @@
+behavior currently.
+
+Note that generating a new name is also a rather expensive operation requiring to traverse the whole
+context, so that's another source of inefficiency.
```
If only we could use `Quote` where it's O(1)...

michaelpj commented on plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs:
```diff
@@ -0,0 +1,312 @@
+-- editorconfig-checker-disable-file
+
+module PlutusCore.Generators.PIR.Substitutions where
```
Maybe we should split out the unification stuff? the substitutions and tools for manipulating them seem potentially independently useful, e.g. could go in `Subst` (making it like the third way of doing substitution oy vey)

effectfully replied:
> Maybe we should split out the unification stuff? 

Done.

> the substitutions and tools for manipulating them seem potentially independently useful, e.g. could go in Subst (making it like the third way of doing substitution oy vey)

Wrote

```
{- Note [Substitution in generators]
Generators define their own substitution algorithm. It's generally useful, however it's quite
inefficient, since calling 'renameVar' in the worker when handling bindings makes the whole
algorithm at least quadratic. Plus, it also very general, since it handles both nested and parallel
substitution.

So for now we've decided to keep it local to the generators and extract it into the main library
once there's a specific reason to do that.
-}
```

Let's keep it in the generators folder for now.

michaelpj commented:
Okay, done one round. I'll try and take a stab at unification later.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+-- | Generate a name that likely overlaps with existing names on purpose. If there are no existing
+-- names, generate a fresh name. This function doesn't distinguish between the type- and term-level
+-- scopes, hence it may generate a 'Name' \"clashing\" with a previously generated 'TyName' and not
+-- clashing with any previously generated 'Name'.
```
No, it's a good thing: we want to check that Plutus is able to handle such spurious name clashes. Generating weird stuff is useful for a testing machinery! We don't need any "definitely non-fresh name" anyway, since we get enough non-fresh names already and trying to add more logic to cover what is the already covered would only needlessly complicate things.

michaelpj replied:
Then maybe say that? As it is it sounds like this is a problem that you haven't fixed, whereas you're arguing that it's a feature! It also wasn't clear that you won't want "definitely non-fresh names", I guess you say that you don't really need them so it's okay, but if you're reliant on the ratio that comes out of this thing in order to make sure you're testing the non-fresh case that you sort-of ask for... seems simpler to just _ensure_ that it does the thing you wan :shrug:

effectfully replied:
> Then maybe say that?

Done.

> but if you're reliant on the ratio

Who cares if it gets skewed slightly.

>  seems simpler to just ensure that it does the thing you wan shrug

Well, the thing is, it's not simpler with the current alignment. Otherwise sure, it'd be cleaner to have `genDefinitelyFreshName`, `genDefinitelyNonFreshName` and `genDefinitelyNonFreshInTheOtherScopeName`. Although it'd still be more logic and code. What we have right now is fine, we don't want to spend months here calculating the exact best possible ratio between fresh and non-fresh names to get a marginal coverage improvement when we can instead look into things on the critical path or indeed improve the generators machinery in places where it matters much more (such as faster fresh name generation, faster substitution, cleverer unification, statistical analysis of generated output etc).

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+  where
+    names (Datatype _ _ _ m cs) = Set.fromList $ _nameUnique m : [ _nameUnique c | VarDecl _ c _ <- cs ]
+
+{- Note [Warning about generating fresh names]: because `GenTm` is a *reader* monad
```
I think the Note is a bit confusing, I'll polish it while I'm here. We can't avoid what the Note describes, because we have to model scopes and for that we need a reader monad and for that we have to have some explicit for of binding variables. The mentioned in the Note overlap thing is different and for that we could indeed use `Quote`, but see PLT-806, at this point it's quite hard to change the logic and I definitely don't want this to be a blocker, giving that the current behavior is not wrong, just not optimal.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+bindTyName :: TyName -> Kind () -> GenTm a -> GenTm a
+bindTyName x k = local $ \ e -> e
+    { geTypes = Map.insert x k (geTypes e)
+    , geTerms = Map.filter (\ty -> not $ x `Set.member` setOf ftvTy ty) (geTerms e)
```
> you're removing things from the environment that match the new varaible?

that reference the now shadowed variable in their type.

michaelpj replied:
Okay, that makes a sort of sense but it seems potentially buggy:
- `geTerms` is _not_ the set of terms that are in scope, but the set of terms in scope that can be correctly inserted into the program. Similarly for the others. This should definitely be documented!
- Fresh name generation relies on the invariant that all the generated uniques appear in one of the in-scope maps, but this invariant is violated by what's happening here, since when you add a fresh type variable you may lose an arbitrary number of term variables and also some constructor variables for any datatypes that get removed.

Or maybe this is fine? If you re-generate one of the removed names you'll just shadow it, and maybe that's fine. But then the use of "fresh name" seems confusing since it means something different to what it means elsewhere.

effectfully replied:
Those are very good points... So `genFreshName` is in fact `getLikelyFreshName`. Definitely worth documenting, I'll do it.

@MaximilianAlgehed so it looks like `genFreshName` can generate a non-fresh name. How do you feel about it, can that break some invariant or something or is it fine?

MaximilianAlgehed replied:
Should be fine, but maybe worth looking in to anyway.

MaximilianAlgehed replied:
I don't mean that it will have bad consequences, but that it might be morally better to have a function that is guaranteed to generate a fresh name.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+bindTyNames = flip $ foldr (uncurry bindTyName)
+
+-- | Remember that we have generated a type name locally but don't bind it.
+-- Useful for non-recursive definitions where we want to control name overlap.
```
This function apparently isn't used anywhere, as well as the `geUnboundUsedTyNames` field of `GenEnv` that it changes (well, it's used, but not in a useful way, I think). I'll remove all of that, thanks for catching.

effectfully replied:
Eeeek... I'm in the types-only part of the generators machinery, maybe the terms part uses it. I should be more careful.

effectfully replied:
Yep, it's used in the terms part. Yeah, I guess, I'll just return it when I start working on the terms part.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+bindTmNames = flip $ foldr (uncurry bindTmName)
+
+-- | Create a fresh term name, bind it to a type, and use it in a generator.
+bindFreshTmName :: String -> Type TyName DefaultUni () -> (Name -> GenTm a) -> GenTm a
```
Yeah, functions that handle bindings always have this kind of interface.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+bindBinds :: Foldable f => f (Binding TyName Name DefaultUni DefaultFun ()) -> GenTm a -> GenTm a
+bindBinds = flip (foldr bindBind)
+
+-- * Containers (zipper-ish, very useful for shrinking.)
```
> Also, I asked before but this definitely isn't on Hackage?

No idea and I think Maximilian didn't have any idea either.

effectfully replied:
> seems independent, put it in its own module?

It's like 20 lines of code, I put the machinery into the `Utils` module.

> Also, I asked before but this definitely isn't on Hackage?

Left the following comment:

```
-- There's https://hackage.haskell.org/package/functor-combo-0.3.6/docs/FunctorCombo-Holey.html
-- (explained in http://conal.net/blog/posts/differentiation-of-higher-order-types), but it's kinda
-- easier to roll out our own version than to dig through the generics in that library. Plus, the
-- library has multiple interfaces looking slightly different and it's not immediately clear how
-- they relate to each other.
```

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs:
```diff
@@ -0,0 +1,176 @@
+  let atoms = [ TyVar () x | (x, k') <- Map.toList tys, k == k' ] ++
+              [ TyVar () x | (x, Datatype _ (TyVarDecl _ _ k') _ _ _) <- Map.toList dts, k == k' ]
+      genBuiltin = fmap (TyBuiltin ()) <$> genBuiltinTypeOf k
+      lam k1 k2 = do
```
There may not be a type variable or a built-in type of the given type, hence we have to resort to generating a lambda occasionally. Plus it's a lambda that ignores the bound variable in its body.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs:
```diff
@@ -0,0 +1,176 @@
+      lam k1 k2 = do
+        x <- genMaybeFreshTyName "a"
+        TyLam () x k1 <$> bindTyName x k1 (genAtomicType k2)
+  -- There's always an atomic type of a given type, hence the usage of 'deliver'.
```
> So it's not really "atomic" types so much as "the smallest type of the given kind that we can come up with".

Smallish, not smallest. We use smallest in the shrinker and there it's called `minimalType`. E.g. for `* -> * -> *` `minimalType` always gives `pair` and `genAtomicType` can give you a type variable, `pair` or a type lambda returning a type variable, `list` or a type lambda returning...

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/ShrinkTypes.hs:
```diff
@@ -0,0 +1,252 @@
+fixKind ctx ty k
+  -- Nothing to do if we already have the right kind
+  | unsafeInferKind ctx ty == k = ty
+  | not $ k `leKind` unsafeInferKind ctx ty = error $ concat
```
Good point, fixed.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/ShrinkTypes.hs:
```diff
@@ -0,0 +1,252 @@
+  list = mkTyBuiltin @_ @[] ()
+  pair = mkTyBuiltin @_ @(,) ()
+
+-- | Take a type in a context and a new target kind
```
Thanks, reworded your rewording a little.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/ShrinkTypes.hs:
```diff
@@ -0,0 +1,252 @@
+              kb'  = unsafeInferKind ctx' b'
+    -- Ill-kinded builtins just go to minimal types
+    TyBuiltin{} -> minimalType k
+    -- Unreachable, because there's no kind that is smaller than @Type ()@.
```
I was confused by that too. If the kind is the same, then we always return the given type back without any adjustments.

michaelpj replied:
Ah okay, then it would be good to explain the reasoning here a bit more: "Unreachable, because the target kind must be less than or equal to the original kind. We handled the case where they are equal at the beginning of the function, so at this point the target kind must be strictly less than the original kind, but for these types we know the original kind is Type, and there is no kind smaller than that."

effectfully replied:
Great suggestion, thanks.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/ShrinkTypes.hs:
```diff
@@ -0,0 +1,252 @@
+    TyForall{} -> error "Internal error: unreachable clause."
+
+-- | Shrink a well-kinded type in a context to new types, possibly with new kinds.
+-- The new kinds are guaranteed to be smaller than or equal to the old kind.
```
Not sure. @MaximilianAlgehed what do you think?

MaximilianAlgehed replied:
Could do, it probably won't have a huge performance impact and might catch a bug or two. I say do it.

michaelpj replied:
Sorry, I didn't mean "check automatically", I just meant "check whether it's actually correct", given that when I wrote this I was suspicious of the previous one.

effectfully replied:
OK, I have no idea what we're discussing here, so I'm blindly resolving this conversation.

[singing "what have I become"]

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/ShrinkTypes.hs:
```diff
@@ -0,0 +1,252 @@
+  -- If we are not already minimal, add the minial type as a possible shrink.
+  [(k, m) | k <- k0 : shrink k0, let m = minimalType k, m /= ty] ++
+  -- TODO: it might be worth-while to refactor this to the structural + nonstructural
+  -- style we use below. Unsure if that's more readable. CODE REVIEW: what do you think?
```
I missed this comment, don't know what it means either. @MaximilianAlgehed do you remember?

MaximilianAlgehed replied:
This refers to the style of shrinking function we wrote for terms where we split the "structural" cases (e.g. shrinking under application) from the "nonstructural" cases (i.e. removing whole subterms all together). The code for that shrinker is more readable than this code so it's probably a decent idea to rewrite this code in that style - but the comment obviously doesn't make much sense any more because it's pointing to code that's not in this PR.

effectfully replied:
OK, I've looked into the term shrinker and I actually find the type shrinker more readable, because you can look at an AST node and observe all the different ways of shrinking it in one place instead of jumping between two, so I'm going to keep the current alignment.

I don't think I'll change the terms alignment either, 'cause it's also perfectly fine. I don't personally care about the slight inconsistency, both the shrinkers are fine.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs:
```diff
@@ -0,0 +1,312 @@
+substitution is passed around in a state monad, since unification results in one part of the
+unification problem are relevant in all the other parts too (this is not in conflict with locals
+being scope-specific, because flexible variable resolution cannot capture a locally bound
+variable). Type unification could take a substitution as an input an extend it instead of always
```
No, it's not `and`, I really meant the article. Maybe it just shouldn't be there? Still can't wrap my head around articles in English.

michaelpj replied:
Er, I edited the wrong "an", I meant to say "substitution as an input _and_ extend it"

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs:
```diff
@@ -0,0 +1,312 @@
+          -> Either String TypeSub
+          -- ^ Either an error or a substitution (from a subset of @flex@) unifying @t1@ and @t2@
+unifyType ctx flex a0 b0 =
+    execStateT (runReaderT (goType (normalizeTy a0) (normalizeTy b0)) Set.empty) Map.empty
```
Because type checking is modulo normalization of types. Hence type unification should also be modulo normalization of types. It's the usual unification business, you can see some examples inline. E.g. we could solve `_f x =?= x -> x` as `_f := \x -> x -> x`.

michaelpj replied:
Right, but what happens if we don't normalize? I guess we just won't notice that some things unify? It sounds like we had some kind of unification of applications, but we probably want to get rid of the ones that actually compute? Can we end up with new computing-applications as we progress through unification? I don't really know what I'm talking about, I should actually read up on unification...

effectfully replied:
> I guess we just won't notice that some things unify?

Yes.

> It sounds like we had some kind of unification of applications, but we probably want to get rid of the ones that actually compute?

Yes, we do absolutely want that. The more complicated your unification machinery is, the more you want to simplify the equation before moving to solving it, particularly since with a complicated unification algorithm you're going to be looking a lot for certain patterns like a metavariable applied to a bunch of variables (a.k.a. Miller pattern unification) etc. I can't imagine any reason to start unification on non-normalized types.

> Can we end up with new computing-applications as we progress through unification?

In the current implementation -- no. In theory -- yes, but it's really complicated.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Subst.hs:
```diff
@@ -145,6 +148,64 @@ termSubstFreeNames
+-- | Substitute a closed 'Type' for a type variable in a 'Type'. Does not descend under binders that
+-- bind the same variable as the one we're substituting for. The resulting 'Term' may and likely
+-- will not satisfy global uniqueness.
+typeSubstClosedType
```
I just removed all `termSubstFreeNames` from the repo. They were kinda weird and we no longer need them, since they were only used for discharging in the CEK machine, which now works in a different way since the change to De Bruijn indices.

effectfully commented on plutus-core/plutus-ir/test/GeneratorSpec/Substitution.hs:
```diff
@@ -0,0 +1,101 @@
+  letCE "rename ty" (runQuote $ rename ty) $ \ rnty ->
+  void $ unifyType mempty mempty ty rnty
+
+-- | Check that substitution gets rid of all the right variables
```
Refined the wording to:

```
-- | Check that substitution eliminates from the type all free occurrences of variables present in
-- the domain of the substitution.
```

I don't want to invent some notation to talk about substitutions just to elaborate on this fairly trivial property.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+
+-- | A type is a container for the purposes of shrinking if it has:
+class Container f where
+  data OneHoleContext f :: * -> *
```
It's fine either way, but `data` is probably a bit more readable compared to using tuples or defining data types outside of the class and hooking them into the class later. Basically, it doesn't really matter, hence I'm resolving the conversation.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenerateKinds.hs:
```diff
@@ -0,0 +1,49 @@
+-- This shrinking order means that `* -> (* -> * -> * -> *) -> *` can shrink to
+-- `* -> *` or `* -> (* -> *) -> *` but not to `* -> * -> * -> *`. Not allowing
+-- this final shrink is important as we are guaranteed to only ever reduce the
+-- number of type arguments we need to provide when shrinking - thus allowing us
```
Definitely. Which is why Maximilian put it in there, I guess :)

How do you spell

```
A kind @k1 = foldr (->) * ks1@ is less than or equal to a kind @k2 = foldr (->) * ks2@ when @ks1@
can be constructed by dropping and shrinking kinds in @ks2@.
```

in Haskell? This way:

```haskell
  -- See Note [Shriking order on kinds].
  shrink (Type _)          = []
  -- Reminder: @a@ can have bigger arity than @a -> b@ so don't shrink to it!
  shrink (KindArrow _ a b) = b : [KindArrow () a' b' | (a', b') <- shrink (a, b)]
```

and the property that you described in this one then:

```haskell
-- | Test that shrinking kinds generates smaller kinds.
prop_shrinkKindSmaller :: Property
prop_shrinkKindSmaller =
  forAllDoc "k" arbitrary shrink $ \ k ->
  assertNoCounterexamples [k' | k' <- shrink k, not $ leKind k' k]
```Definitely. Which is why Maximilian put it in there, I guess :)

How do you spell

```
A kind @k1 = foldr (->) * ks1@ is less than or equal to a kind @k2 = foldr (->) * ks2@ when @ks1@
can be constructed by dropping and shrinking kinds in @ks2@.
```

in Haskell? This way:

```haskell
  -- See Note [Shriking order on kinds].
  shrink (Type _)          = []
  -- Reminder: @a@ can have bigger arity than @a -> b@ so don't shrink to it!
  shrink (KindArrow _ a b) = b : [KindArrow () a' b' | (a', b') <- shrink (a, b)]
```

and the property that you described in this one then:

```haskell
-- | Test that shrinking kinds generates smaller kinds.
prop_shrinkKindSmaller :: Property
prop_shrinkKindSmaller =
  forAllDoc "k" arbitrary shrink $ \ k ->
  assertNoCounterexamples [k' | k' <- shrink k, not $ leKind k' k]
```

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+           Set.mapMonotonic _nameUnique (Map.keysSet tms) <>
+           Set.unions [ names d | d <- Map.elems dts ]
+  where
+    names (Datatype _ _ _ m cs) = Set.fromList $ _nameUnique m : [ _nameUnique c | VarDecl _ c _ <- cs ]
```
> hmm, do I not have this somewhere... apparently not, probably could add some more Traversals for stuff like this if we wanted.

It's a one-liner here, so I don't personally care.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs:
```diff
@@ -0,0 +1,176 @@
+  tys <- asks geTypes
+  dts <- asks geDatas
+  let atoms = [ TyVar () x | (x, k') <- Map.toList tys, k == k' ] ++
+              [ TyVar () x | (x, Datatype _ (TyVarDecl _ _ k') _ _ _) <- Map.toList dts, k == k' ]
```
> Hmmm, this makes me think that the datas map is weird. It maps the datatype name to the datatype, but the datatype introduces multiple names. I guess it's fine to use that as a handle, though.

Left this comment in the code, not gonna do anything about it.

effectfully commented on plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs:
```diff
@@ -0,0 +1,317 @@
+
+-- * Dealing with size
+
+-- | Map a function over the generator size
```
> I read this and I was immediately confused. I think it's maybe a mistake to use the same name "size" for both the quickcheck size and geSize. Could we call the recursion size "recursion depth" or something? Otherwise I'm going to get confused every time I see one of these size functions.

Depth is a very different thing. E.g. one of the major differences between NEAT and SmallCheck is that the former uses size and the latter uses depth. I get your point though, made it `*AstSize` instead of `*Size` everywhere.

effectfully commented:
I've addressed all the comments and I'm hitting merge, despite some minorish unresolved issues (for which there are Jira tickets). In particular the ones discussed in the Note I added last:

```
{- Note [Chaotic Good fresh name generation]
Currently we don't use 'Quote' for generating fresh names in the QuickCheck generators. It's mostly
for historical reasons and it's very non-trivial to change the current alignment to use 'Quote' at
this point. It does have positive sides to it, though, which we'll discuss below.

So what we do right now is track all names that are in the scope (not all actually and not in the
scope, more on that later, for now let's pretend the statement is correct) in separate maps:

- one for type names
- one for term names
- one for datatype-associated names (the name of the data type, its matcher and constructors)

and all these maps are processed in a linear fashion each time we need to generate a bunch of fresh
names. The processing is done via collecting in a set all uniques of names found in the maps: the
key sets of all the three maps and the names in the elements of the data map. Why do we need the
latter given that the names there are present as keys in the term map? No idea.

What makes it hard to change the current alignment to use 'Quote' is that names are generated all
over the place and put into the current context, inside and outside of the 'GenTm' monad. Any
operation on the context automatically makes the name generation machinery aware of the changes,
because the machinery looks at the whole context each time. It's really hard to move all of that to
run in 'Quote' simply because it’s trivial to miss a call here and there and types don't help much.

Another challenge is that the current fresh name generation is pretty fancy and doesn’t just always
pick the next fresh name. It can pick the next 10th fresh name, then go back, then again forward etc
– all of that between different calls to the function generating fresh names. We want to keep this
logic as it allows us to produce more diverse terms. We probably can simulate this kind of logic
with 'Quote', but it's non-trivial for sure.

Finally, we do not have any function that generates a definitely fresh name or a function that
generates a definitely non-fresh name -- both for very technical reasons (see the docs inline).
This likely isn't intentional, but it's hard to change that at this point and we don't care much
anyway, name clashes are fine. Moreover, the term map rather than containing all term names that
are in the scope only contains those that can be correctly inserted into the program (i.e. don't
reference shadowed type variables in their type). However the data map isn't filtered the same way
as the term map and so there can be constructors whose type references shadowed type variables,
which we never notice because those are only used for fresh name generation... which I guess is how
looking at the elements of the data map isn't redundant in the end.

Overall, this chaotic goodness needs to be sorted out.
-}
```

I'm going to make a new law: time spent resolving issues related to name handling exceeds time spent resolving all other types of issues combined.

effectfully commented:
@michaelpj what's up with CI? Could you merge the PR manually?

michaelpj commented:
Ugh, that again. Very annoying. Yes, I'll force merge.

effectfully commented:
YAY
